### ISSUE TITLE
[Bug]: Tag not found when repo is packed

### ISSUE THREAD

**Issue by thisirs** [progfolio/elpaca#171]
### Elpaca Version

((elpaca . "2b4be7a HEAD -> master, origin/master, origin/HEAD")
 (isntaller . 0.5)
 (emacs .
        "GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.37, cairo version 1.16.0)\n of 2023-06-29")
 (git . "git version 2.40.1"))

### Operating System

Archlinux

### Description

Hi,

I am not able to check out a specific tag. Apparently, a tag can be found under `.git/refs/tags/` unless the repo is packed. In that case, `git checkout .git/refs/tags/0.0.7` fails. One possible fix is to use `git checkout tags/0.0.7` in `elpaca--checkout-ref`. WDYT ?

**Comment by progfolio** [comment]
Thanks for taking the time to fill out a bug report.
Could you please be more specific?
It would be helpful to know what repo specifically, and see a backtrace or error log of how it is failing.
Thanks.

**Comment by progfolio** [comment]
>One possible fix is to use `git checkout tags/0.0.7` in `elpaca--checkout-ref`. WDYT ?

I see what you mean now. This stands to reason.
I've pushed a fix which should take care of that.
Testing is appreciated. Thank you.

**Comment by progfolio** [comment]
Seems to work fine on my end.
One thing to note is that you'll want to ensure the clone is not shallow.
Otherwise, the tags will not be fetched.
You can do that by passing `:depth nil` in the recipe.

Ideally, we could handle this case (and others, such as when a specific commit is given via the `:ref` recipe keyword) by recursively fetching until the ref we're after is present.
That way, you wouldn't need to specify the `:depth` in those cases.
I'll have to think on that more before I decide on whether to implement it or not.

<details open><summary>Test Case</summary>

```emacs-lisp
(elpaca-test
  :init (elpaca (ement :depth nil :tag "v0.9.1")
          (elpaca-with-dir 'ement repo
            (print (elpaca-process-output "git" "status"))))
  (elpaca-wait))
```

</details>
<details><summary>Host Env</summary>

<table>
<tr><td>elpaca</td><td>a54ce46 HEAD -> master, origin/master, origin/HEAD</td>
<tr><td>isntaller</td><td>0.5</td>
<tr><td>emacs</td><td>GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8) of 2023-07-25</td>
<tr><td>git</td><td>git version 2.41.0</td>
</table>
</details>

<details><summary>Output</summary>

```emacs-lisp
  INFO     Scraping files for loaddefs... 
  INFO     Scraping files for loaddefs...done
  GEN      ../elpaca-autoloads.el
Cloning into '/tmp/elpaca.v9IpDn/elpaca/repos/elpaca'...
Your branch is up to date with 'origin/master'.
Checking /tmp/elpaca.v9IpDn/elpaca/repos/elpaca...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-info.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-log.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-manager.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-menu-elpa.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-menu-melpa.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-menu-org.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-process.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-test.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca-ui.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/elpaca.el...
Checking /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/doc...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/doc/early-init.el...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/doc/init.el...
Checking /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/extensions...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/extensions/elpaca-use-package.el...
Checking /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/images...
Checking /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/test...
Compiling /tmp/elpaca.v9IpDn/elpaca/repos/elpaca/test/elpaca-test.el...
Done (Total of 11 files compiled, 3 skipped in 4 directories)

Downloading MELPA recipes...
Downloading MELPA recipes...100%
Downloading NonGNU-devel ELPA...
Downloading GNU-devel ELPA...
Downloading NonGNU ELPA...
Downloading GNU ELPA...

"HEAD detached at v0.9.1
nothing to commit, working tree clean
"


 Test Env

Elpaca a54ce46 HEAD -> master, origin/master, origin/HEAD
installer:      0.5
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8)
 of 2023-07-25
git --version:  git version 2.41.0
```
</details>
